src/challenges.js

```diff
@@ -1,45 +1,110 @@
 /* ======================== CallBacks Practice ============================ */
 const each = (elements, cb) => {
-  // Iterates over a list of elements, yielding each in turn to the `cb` function.
-  // This only needs to work with arrays.
+  for (let i = 0; i < elements.length; i++) {
+    cb(elements[i], i);
+  }
 };
 
 const map = (elements, cb) => {
-  // Produces a new array of values by mapping each value in list through a transformation function (iteratee).
-  // Return the new array.
+  const retArr = [];
+  each(elements, (element, i) => {
+    retArr[i] = cb(element);
+  })
+  return retArr;
 };
 
 /* ======================== Closure Practice ============================ */
 const limitFunctionCallCount = (cb, n) => {
-  // Should return a function that invokes `cb`.
-  // The returned function should only allow `cb` to be invoked `n` times.
+  let count = 0;
+  return (...args) => {
+    if (count < n) {
+      count++;
+      return cb(...args);
+    }
+    return null;
+  };
 };
 
 const cacheFunction = cb => {
-  // Should return a funciton that invokes `cb`.
-  // A cache (object) should be kept in closure scope.
-  // The cache should keep track of all arguments have been used to invoke this function.
-  // If the returned function is invoked with arguments that it has already seen
-  // then it should return the cached result and not invoke `cb` again.
-  // `cb` should only ever be invoked once for a given set of arguments.
+  const cacheObj = {};
+  return (...args) => {
+    const key = args;
+    if (!(key in cacheObj)) {
+      const cacheVal = cb(...args);
+      cacheObj[key] = cacheVal;
+      return cacheVal;
+    }
+    return cacheObj.key;
+  };
 };
 
 /* eslint-enable no-unused-vars */
 
 /* ======================== Recursion Practice ============================ */
 const reverseStr = str => {
-  // reverse str takes in a string and returns that string in reversed order
-  // The only difference between the way you've solved this before and now is that you need to do it recursivley!
+  if (str.length == 1) return str;
+  return str[str.length - 1] + reverseStr(str.slice(0, str.length - 1));
 };
 
 const checkMatchingLeaves = obj => {
-  // return true if every property on `obj` is the same
-  // otherwise return false
+  // recur base case when there are only two properties in Object
+  if (Object.values(obj).length === 2) {
+    const objValsIndexZero = Object.values(obj)[0];
+    const objValsIndexOne = Object.values(obj)[1];
+    // check if first and second properties are comparable (index 0 and 1 resp.)
+    // comparable is defined here as an object property with
+    // 1) no Object nesting and
+    // 2) no Array nesting
+    // so if an Object's properties is NOT an Object and NOT an array
+    // these properties are comparable using ===
+    if (!(typeof objValsIndexZero === 'object') && !Array.isArray(objValsIndexZero)) {
```

Comment:
while this might be a good use case to check for an array as well. We were only ever supplying this algorithm with an object of nested objects. You dind't need this check.